packages/core-mobile/app/new/routes/_layout.tsx

```diff
@@ -56,32 +59,47 @@ export default function RootLayout(): JSX.Element | null {
 
   useLoadFonts()
 
+  // please be careful with the dependencies here
+  // this effect is responsible for redirecting users
+  // to either the sign up flow or the login modal
+  // we don't want to include any other dependencies here
   useEffect(() => {
     if (walletState === WalletState.NONEXISTENT) {
       if (router.canGoBack()) {
         navigation.dispatch(StackActions.popToTop())
       }
       router.replace('/signup')
     } else if (walletState === WalletState.INACTIVE) {
-      // Navigate to login modal when walletState is not active
+      // navigate to login modal when wallet is not active
       router.navigate('/loginWithPinOrBiometry')
-    } else if (walletState === WalletState.ACTIVE) {
-      if (canGoBackToWallet && router.canGoBack()) {
-        router.back()
-      } else {
-        router.replace('/portfolio')
-      }
     }
-  }, [walletState, router, canGoBackToWallet, navigation])
+  }, [walletState, router, navigation])
 
   useEffect(() => {
-    // set shouldRenderOnlyPinScreen to false once wallet is unlocked
-    // do nothing if app is not ready (as we need to sync wallet state after rehydration)
-    // or if we have already set shouldRenderOnlyPinScreen to false
-    if (!appIsReady || shouldRenderOnlyPinScreen === false) return
+    /**
+     * after the wallet is successfully unlocked
+     *
+     * - redirect to the portfolio if:
+     *   - the app was freshly opened
+     *   - the user just completed onboarding
+     * - otherwise, return the user to their previous/last screen if the app was locked due to inactivity.
+     */
+    if (walletState === WalletState.ACTIVE) {
+      // when the login modal is the last route and on top of the (signedIn) stack
+      // it means the app just resumed from inactivity
+      const isReturningFromInactivity =
+        isSignedIn && pathName === '/loginWithPinOrBiometry'
 
-    setShouldRenderOnlyPinScreen(walletState !== WalletState.ACTIVE)
-  }, [appIsReady, shouldRenderOnlyPinScreen, walletState])
+      if (isReturningFromInactivity) {
+        router.canGoBack() && router.back()
+      } else if (
+        pathName === '/onboarding/mnemonic/confirmation' || // onboarding completion
+        (pathName === '/loginWithPinOrBiometry' && !isSignedIn) // fresh app open
+      ) {
+        router.replace('/portfolio')
```

Comment:
this was the main cause of the triple rendering of portfolio. we were calling `router.replace('/portfolio')` repeatedly. I added some checks here to make sure we only call it when needed.